### PR TITLE
Savotina Valeria, Lab №1, var: 2.

### PR DESCRIPTION
**Pull request by Savotina**
The plugin allows all functions that don't have conditions (for example, if, do...while, etc.), add the `always_inline` attribute.

### CODE REVIEW

File: clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+class AddAttrAlwaysInlineAction final : public clang::PluginASTAction {
+public:
+  std::unique_ptr<clang::ASTConsumer>
+  CreateASTConsumer(clang::CompilerInstance &ci, llvm::StringRef) override {
+    return std::make_unique<AddAttrAlwaysInlineConsumer>(&ci.getASTContext());
+  }
+
+  bool ParseArgs(const clang::CompilerInstance &ci,
+                 const std::vector<std::string> &args) override {
+    return true;
```
Comment (m-ly4) [inline clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp:88]: Could you please add `help` option with descriptive output and a test for it?

File: clang/test/lab1/savotina_valeria/tests.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/AddAttrAlwaysInlinePlugin%pluginext\
+// RUN: -plugin add_attr_always_inline_plugin %s 1>&1 | FileCheck %s
```
Comment (m-ly4) [inline clang/test/lab1/savotina_valeria/tests.cpp]: Could you please check attribute presence via searching in AST dump output?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          if (!findСonditionalStatement(body)) {
+            if (auto loc = decl->getSourceRange(); loc.isValid()) {
+              decl->addAttr(
+                  clang::AlwaysInlineAttr::Create(*context, loc.getBegin()));
+            }
+          }
+        }
+      }
+      outStatusAttrAlwaysInline(decl);
+      llvm::outs() << "==================================\n";
```
Comment (aobolensk) [inline clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp]:
```suggestion
```
It looks like debug output, please remove such things

Comment (Savotina) [reply clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp]: @m-ly4 @aobolensk I've corrected what you asked for

File: clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        if (auto body = decl->getBody()) {
+          if (!findСonditionalStatement(body)) {
+            if (auto loc = decl->getSourceRange(); loc.isValid()) {
+              decl->addAttr(
+                  clang::AlwaysInlineAttr::Create(*context, loc.getBegin()));
+            }
+          }
+        }
+      }
+    }
```
Comment (m-ly4) [inline clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp]: In LLVM it is customary to reduce nesting as much as possible:
```suggestion
    if (!decl->isFunctionOrFunctionTemplate())
      return true;
    if (decl->hasAttr<clang::AlwaysInlineAttr>())
      return true;
    if (auto body = decl->getBody()) {
      if (findСonditionalStatement(body))
        return true;

      if (auto loc = decl->getSourceRange(); loc.isValid())
        decl->addAttr(
            clang::AlwaysInlineAttr::Create(*context, loc.getBegin()));
    }
```

File: clang/test/lab1/savotina_valeria/tests.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// RUN: %clang_cc1 -load %llvmshlibdir/AddAttrAlwaysInlinePlugin%pluginext\
```
Comment (m-ly4) [inline clang/test/lab1/savotina_valeria/tests.cpp:1]: @Savotina  Test for `help`?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp (hunk 4 above)
Comment (Savotina) [reply clang/labs/lab1/savotina_valeria/AddAttrAlwaysInline.cpp]: @m-ly4 ok, I see, thanks!

File: clang/test/lab1/savotina_valeria/tests.cpp (hunk 5 above)
Comment (Savotina) [reply clang/test/lab1/savotina_valeria/tests.cpp:1]: @m-ly4 done

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**